openstack/datadog_checks/openstack/openstack.py

```diff
@@ -936,41 +934,69 @@ def get_all_servers(self, i_key, collect_all_tenants, filter_by_host=None):
+                if new_server['project_name']:
+                    self.server_details_by_id[new_server['server_id']] = new_server
             elif new_server['server_id'] in self.server_details_by_id and new_server['state'] in REMOVED_STATES:
                 self.log.debug("Removing server from cache: %s", new_server)
                 try:
                     del self.server_details_by_id[new_server['server_id']]
                 except KeyError as e:
                     self.log.debug("Server: %s has already been removed from the cache", new_server['server_id'])
 
-        return self.server_details_by_id
+    def filter_excluded_servers(self):
+        proj_list = set([])
+        if self.exclude_server_id_rules:
+            # Filter out excluded servers
+            for exclude_id_rule in self.exclude_server_id_rules:
+                for server_id in self.server_details_by_id.keys():
+                    if re.match(exclude_id_rule, server_id):
+                        del self.server_details_by_id[server_id]
+
+        for _, server in self.server_details_by_id.iteritems():
+            proj_list.add(server.get('project_name'))
+
+        projects_filtered = pattern_filter(
+            proj_list,
+            whitelist=self.include_project_name_rules,
+            blacklist=self.exclude_project_name_rules
+        )
+
+        self.server_details_by_id = {
+                sid: server for (sid, server)
+                in self.server_details_by_id.iteritems()
+                if server.get('project_name') in projects_filtered
+        }
 
     def get_project_name_from_id(self, tenant_id):
-        url = "{0}/{1}/{2}/{3}".format(self.keystone_server_url, DEFAULT_KEYSTONE_API_VERSION, "projects", tenant_id)
+        url = "{}/{}/{}/{}".format(self.keystone_server_url, DEFAULT_KEYSTONE_API_VERSION, "projects", tenant_id)
         self.log.debug("Project URL is %s", url)
         headers = {'X-Auth-Token': self.get_auth_token()}
         try:
             r = self._make_request_with_auth_fallback(url, headers)
             return r['project']['name']
 
         except Exception as e:
-            self.warning('Unable to get project name: {0}'.format(str(e)))
-            raise e
+            self.warning('Unable to get project name: {}'.format(str(e)))
+            return None
 
-    def get_stats_for_single_server(self, server_details, tags=None):
+    def get_stats_for_single_server(self, server_details, tags=None, use_shortname=False):
         def _is_valid_metric(label):
             return label in NOVA_SERVER_METRICS or any(seg in label for seg in NOVA_SERVER_INTERFACE_SEGMENTS)
 
+        hypervisor_hostname = server_details.get('hypervisor_hostname')
+        host_tags = self._get_host_aggregate_tag(hypervisor_hostname, use_shortname=use_shortname)
+        host_tags.append('availability_zone:{}'.format(server_details.get('availability_zone', 'NA')))
+
```

Comment:
looks like we should add `self.external_host_tags[server] = host_tags` here in order to populate metadatas

Reply:
I am not sure if it should include the `availability_zone` tag tho. I don't see why not but it doesn't seem it was added before

Reply:
It looks to have been a host tag before - https://github.com/DataDog/integrations-core/blob/master/openstack/datadog_checks/openstack/openstack.py#L1405